lib/elixir/lib/uri.ex

```diff
@@ -1021,6 +1021,113 @@ defmodule URI do
   def append_path(%URI{}, path) when is_binary(path) do
     raise ArgumentError, ~s|path must start with "/", got: #{inspect(path)}|
   end
+
+  @doc """
+  Normalizes the given `uri` following the rules defined by RFC 3986.
+
+  ## Examples
+
+      iex> URI.normalize(URI.parse("HtTp://example.com/")) |> URI.to_string()
+      "http://example.com/"
+
+      iex> URI.normalize(URI.parse("http://ExAmPlE.CoM/")) |> URI.to_string()
+      "http://example.com/"
+
+      iex> URI.normalize(URI.parse("http://example.com/%61")) |> URI.to_string()
+      "http://example.com/a"
+
+      iex> URI.normalize(URI.parse("http://example.com/%7Ejane")) |> URI.to_string()
+      "http://example.com/~jane"
+
+      iex> URI.normalize(URI.parse("http://example.com/?q=1%2f2")) |> URI.to_string()
+      "http://example.com/?q=1%2F2"
+
+      iex> URI.normalize(URI.parse("http://example.com/a/./b")) |> URI.to_string()
+      "http://example.com/a/b"
+
+      iex> URI.normalize(URI.parse("http://example.com/./a/b/../c/./")) |> URI.to_string()
+      "http://example.com/a/c/"
+
+      iex> URI.normalize(URI.parse("http://example.com")) |> URI.to_string()
+      "http://example.com/"
+
+      iex> URI.normalize(URI.parse("http://example.com:/")) |> URI.to_string()
+      "http://example.com/"
+
+      iex> URI.normalize(URI.parse("http://example.com:80/")) |> URI.to_string()
+      "http://example.com/"
+
+      iex> URI.normalize(URI.parse("https://example.com:443/")) |> URI.to_string()
+      "https://example.com/"
```

Comment:
We have too many examples, we should probably move some of these to tests and minimize the examples here. :)